pkg/storage/ingest/partition_offset_client.go

```diff
@@ -238,3 +249,136 @@ func (p *partitionOffsetClient) FetchTopicLastProducedOffsets(ctx context.Contex
+	shards := p.client.RequestSharded(ctx, req)
+
+	for _, shard := range shards {
+		if shard.Err != nil {
+			return nil, shard.Err
+		}
+
+		res := shard.Resp.(*kmsg.ListOffsetsResponse)
+		if len(res.Topics) != 1 {
+			return nil, fmt.Errorf("unexpected number of topics in the response (expected: %d, got: %d)", 1, len(res.Topics))
+		}
+		if res.Topics[0].Topic != p.topic {
+			return nil, fmt.Errorf("unexpected topic in the response (expected: %s, got: %s)", p.topic, res.Topics[0].Topic)
+		}
+
+		for _, pt := range res.Topics[0].Partitions {
+			if err := kerr.ErrorForCode(pt.ErrorCode); err != nil {
+				return nil, err
+			}
+
+			list[p.topic][pt.Partition] = kadm.ListedOffset{
+				Topic:       p.topic,
+				Partition:   pt.Partition,
+				Timestamp:   pt.Timestamp,
+				Offset:      pt.Offset,
+				LeaderEpoch: pt.LeaderEpoch,
+			}
+		}
+	}
+
+	return list, nil
+}
+
+// listTopicPartitionIDs returns a list of all partition IDs in the topic.
+func (p *partitionOffsetClient) listTopicPartitionIDs(ctx context.Context) ([]int32, error) {
+	topics, err := p.admin.ListTopics(ctx, p.topic)
+	if err != nil {
+		return nil, err
+	}
+	if err := topics.Error(); err != nil {
+		return nil, err
+	}
+
+	// We expect 1 topic in the response.
+	if len(topics) != 1 {
+		return nil, fmt.Errorf("unexpected number of topics in the response (expected: %d, got: %d)", 1, len(topics))
+	}
+
+	// Get the expected topic.
+	topic, ok := topics[p.topic]
+	if !ok {
+		return nil, fmt.Errorf("unexpected topic in the response (expected: %s)", p.topic)
+	}
+
+	// Extract the partition IDs from the response.
+	ids := make([]int32, 0, len(topic.Partitions))
+	for _, partition := range topic.Partitions {
+		// We don't check partition.Err because all we need is the partition ID;
+		// kadm.Client.ListEndOffsets() doesn't check it neither.
+		ids = append(ids, partition.Partition)
+	}
+
+	// It's not required to sort the partition IDs, but we do it to simplify tests and troubleshooting.
+	slices.Sort(ids)
+
+	return ids, nil
+}
+
+// listTopicPartitionIDsWithCache returns a list of all partition IDs in the topic. The returned list
+// of partition IDs may be picked up from the in-memory cache if the previously cached list has been
+// refreshed less than ttl time ago.
+func (p *partitionOffsetClient) listTopicPartitionIDsWithCache(ctx context.Context, ttl time.Duration) ([]int32, error) {
+	// Lookup cache.
+	p.topicPartitionIDsMx.RLock()
+	var cachedPartitionIDs []int32
+	if !p.topicPartitionIDsUpdatedAt.IsZero() && time.Since(p.topicPartitionIDsUpdatedAt) <= ttl {
+		cachedPartitionIDs = slices.Clone(p.topicPartitionIDs)
+	}
+	p.topicPartitionIDsMx.RUnlock()
+
+	if cachedPartitionIDs != nil {
+		return cachedPartitionIDs, nil
+	}
+
+	// Fetch partition IDs.
+	partitionIDs, err := p.listTopicPartitionIDs(ctx)
+	if err != nil {
+		return partitionIDs, err
+	}
+
+	// Cache it.
+	p.topicPartitionIDsMx.Lock()
+	p.topicPartitionIDs = slices.Clone(partitionIDs)
```

Comment:
[nit] Do we need a clone here? The return type from `listTopicPartitionIDs` are a slice of `[]int32` anyways.

Reply:
I'm cloning it, so that if the `listTopicPartitionIDsWithCache()` caller modifies the returned `[]int32` then the cached version is not modified too.